### PR TITLE
Use function signatures for SILDeclRefs in witness_tables, vtables and witness_method instructions

### PR DESCRIPTION
**Pull request by swiftix**
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

A smoke test on macOS does the following:
1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:
1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

Textual SIL was sometimes ambiguous when SILDeclRefs were used, because the textual representation of SILDeclRefs was the same for functions that have the same name, but different signatures.

This patch fixes this problem.

### CODE REVIEW

**Comment by swiftix** [comment]
@swift-ci Please smoke test

**Comment by swiftix** [comment]
@jckarter Joe, do you mind having a quick look? This is nothing urgent. So, only if you have a couple of spare minutes.

**Comment by gottesmm** [comment]
@swiftix Is this doing the demangled thing?

**Comment by swiftix** [comment]
@gottesmm No, it adds signatures. So, no cryptic mangled names.

**Comment by gottesmm** [comment]
= D

**Comment by swiftix** [comment]
@swift-ci Please test

**Comment by swiftix** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
Looks good, thanks @swiftix .

**Comment by shahmishal** [comment]
@swift-ci please test linux

**Comment by tkremenek** [comment]
Linux failure looks genuine.

**Comment by swiftix** [comment]
@swift-ci Please smoke test Linux

**Comment by swiftix** [comment]
There seems to be a problem with parsing SILDeclRefs which use a Self type. I filed a rdar://27735857 about it.

@jckarter @slavapestov Can one of you have a look at this radar when you have a spare minute?

**Comment by swiftix** [comment]
This PR is dependent on the #4126.

**Comment by slavapestov** [comment]
@swiftix The DynamicSelf fix is now merged in master. You don't need it for 3.0 branch do you?

**Comment by swiftix** [comment]
@slavapestov No, I don't need to for 3.0.

**Comment by swiftix** [comment]
@swift-ci Please test

**Comment by swiftix** [comment]
@swift-ci Please smoke test

**Comment by swiftix** [comment]
@swift-ci Please test

**Comment by swiftix** [comment]
@swift-ci Please test Linux

**Comment by swiftix** [comment]
@swift-ci Please test Linux

**Comment by swiftix** [comment]
@swift-ci Please test Linux

**Comment by swiftix** [comment]
@swift-ci Please test Linux

**Comment by swiftix** [comment]
@akyrtzi I'm blocked by rdar://27794707 now ;-) If you could fix it, it would be nice. But it is not urgent. So, no hurry.

**Comment by tkremenek** [comment]
@swiftix  rdar://27794707 is fixed

**Comment by slavapestov** [comment]
@swiftix Just saying, it would be awesome if you ever get around to finishing this off :-)

**Comment by swiftix** [comment]
@swift-ci Please smoke test

**Comment by swiftix** [comment]
@swift-ci Please smoke test

**Comment by shahmishal** [comment]
FYI - CI is down for upgrade.

**Comment by swiftix** [comment]
@jckarter Do you mind taking a quick look at it? Does it seem reasonable?

**Comment by swiftix** [comment]
@jckarter I'd like to merge it today. Any objections?

File: test/SILGen/dynamic_lookup.swift [hunk 1]
```diff
@@ -58,7 +58,7 @@ func direct_to_static_method(_ obj: AnyObject) {
   // CHECK-NEXT: [[OBJCOPY:%[0-9]+]] = load_borrow [[PBOBJ]] : $*AnyObject
   // CHECK-NEXT: [[OBJMETA:%[0-9]+]] = existential_metatype $@thick AnyObject.Type, [[OBJCOPY]] : $AnyObject
   // CHECK-NEXT: [[OPENMETA:%[0-9]+]] = open_existential_metatype [[OBJMETA]] : $@thick AnyObject.Type to $@thick (@opened([[UUID:".*"]]) AnyObject).Type
-  // CHECK-NEXT: [[METHOD:%[0-9]+]] = dynamic_method [volatile] [[OPENMETA]] : $@thick (@opened([[UUID]]) AnyObject).Type, #X.staticF!1.foreign : (X.Type) -> () -> (), $@convention(objc_method) (@thick (@opened([[UUID]]) AnyObject).Type) -> ()
+  // CHECK-NEXT: [[METHOD:%[0-9]+]] = dynamic_method [volatile] [[OPENMETA]] : $@thick (@opened([[UUID]]) AnyObject).Type, #X.staticF!1.foreign : (X.Type) -> () -> () , $@convention(objc_method) (@thick (@opened([[UUID]]) AnyObject).Type) -> ()
```
Comment (jckarter) [inline test/SILGen/dynamic_lookup.swift]: Extra space seems to be printed here

**Comment by jckarter** [comment]
It'd be nice to have a test with a `.sil` file that exercises the new syntax. Looks reasonable otherwise.

**Comment by gottesmm** [comment]
@jckarter We definitely need a .sil file test to ensure that the parser can correctly handle this. I would also be fine with round tripping *.swift through sil-opt.

**Comment by gottesmm** [comment]
But a *.sil file is a better solution since it creates a mark in the codebase that this code is tested in an expected way.

**Comment by swiftix** [comment]
@swift-ci Please smoke test

**Comment by swiftix** [comment]
@swift-ci Please smoke test

**Pull request merged**